Source/modules/webcl/WebCLCommandQueue.cpp

```diff
@@ -0,0 +1,1301 @@
+
+    cl_event* clEventId = WebCLEventPtrToCLEventPtr(event, es);
+    if (event && !clEventId)
+        return;
+
+    cl_int err = clEnqueueCopyBufferToImage(m_clCommandQueue, clSrcBufferId, clDstImageId, offset, targetOriginCopy.data(), regionCopy.data(), clEvents.size(), clEvents.data(), clEventId);
+    if (err != CL_SUCCESS)
+        WebCLException::throwException(err, es);
+}
+
+WebCLCommandQueue::WebCLCommandQueue(cl_command_queue commandQueue, WebCLContext* context, WebCLDevice* device)
+    : ThreadSafeWebCLObject(context)
+    , m_callback(nullptr)
+    , m_eventForCallback(nullptr)
+    , m_device(device)
+    , m_isQueueBlockedByFinish(false)
+    , m_clCommandQueue(commandQueue)
+{
+}
+
+Vector<cl_event> WebCLCommandQueue::WebCLEventVectorToCLEventVector(bool blocking, Vector<RefPtr<WebCLEvent>> events, ExceptionState& es)
+{
+    Vector<cl_event> clEvents;
+    for (auto event : events) {
+        if (event->isReleased()) {
+            es.throwWebCLException(WebCLException::INVALID_EVENT_WAIT_LIST, WebCLException::invalidEventWaitListMessage);
+            break;
+        }
+        if (!WebCLInputChecker::compareContext(context(), event->context())) {
+            es.throwWebCLException(WebCLException::INVALID_CONTEXT, WebCLException::invalidContextMessage);
+            break;
+        }
+        if (blocking && events[0]->getStatus() < 0) {
+            es.throwWebCLException(WebCLException::EXEC_STATUS_ERROR_FOR_EVENTS_IN_WAIT_LIST, WebCLException::execStatusErrorForEventsInWaitListMessage);
+            break;
+        }
+        if (blocking && (event->isUserEvent() || !event->setAssociatedCommandQueue(this))) {
+            es.throwWebCLException(WebCLException::INVALID_EVENT_WAIT_LIST, WebCLException::invalidEventWaitListMessage);
+            break;
+        }
+
+        clEvents.append(event->getEvent());
+    }
+    return clEvents;
+}
+
+cl_event* WebCLCommandQueue::WebCLEventPtrToCLEventPtr(WebCLEvent* event, ExceptionState& es)
+{
+    cl_event* clEventId = nullptr;
+    if (event) {
+        clEventId = &event->platformObjectRef();
+        if (!clEventId || !event->setAssociatedCommandQueue(this)) {
+            es.throwWebCLException(WebCLException::INVALID_EVENT, WebCLException::invalidEventMessage);
+            return nullptr;
+        }
+    }
+    return clEventId;
+}
+
+bool WebCLCommandQueue::isExtensionEnabled(WebCLContext* context, const String& name) const
+{
+    return context->isExtensionEnabled(name);
+}
+
+void WebCLCommandQueue::callbackProxy(cl_event event, int type, void* userData)
+{
+    WebCLCommandQueue* webCommandQueue = static_cast<WebCLCommandQueue*>(userData);
+    cl_event* eventForCallback = nullptr;
+    if (webCommandQueue)
+        eventForCallback = webCommandQueue->eventForCallback();
+
+    if (type != CL_COMPLETE || event != *eventForCallback)
+        return;
+
+    if (!isMainThread()) {
+        callOnMainThread(WTF::bind(callbackProxyOnMainThread, webCommandQueue));
```

Comment:
commandQueue can be deleted between this call and callbackProxyOnMainThread execution. How do you make sure commandQueue lifetime?